### PR TITLE
Closes #44 global object for spec file

### LINKED ISSUE #44: Feature Request:  ?global? object for spec file
## Background

 IS there a way to make my `adds_spec` file into some type of object that the `xportr_` functions can access without have to call it over and over again.
 
Have:
```
 adds %>% 
  select(adds_spec$variable) %>% 
  xportr_order(adds_spec) %>% 
  xportr_label(adds_spec) %>% 
  xportr_format(adds_spec)
```

want:
```
 adds %>% 
  select(adds_spec$variable) %>% 
  xportr_order() %>% 
  xportr_label() %>% 
  xportr_format()
```

I've seen you all do some skulduggery awesomeness with R so wondering if this is something we can set up for users or show users how to do in an articel.  @elimillera  

## Proposal

See https://github.com/atorus-research/xportr/issues/16#issuecomment-1531724342

We discussed folding in an additional aspect with this issue: reducing duplication of arguments in a pipeline. Right now, the metadata and possibly the domain must be specified at every step of a pipeline. The goal I have in mind is to be able to run something like this:

adsl_data %>%
  xportr_set(metadata_obj, optional_domain) %>%
  xportr_type() %>%
  xportr_length() %>%
  ...etc.`
  
##  Questions
  
- [ ] How to reduce burden on user?
- [ ] Warnings and Errors when not used properly

## Definition of Done
- [ ] Create function that reduces burden of repetitive entries in xportr functions
- [ ] Downstream functions are not impacted, i.e. `xportr_` functions can be used without this function
- [ ] Unit Tests created and documentation written

### PR DESCRIPTION
**Pull request by EeethB**
### Thank you for your Pull Request! 

We have developed a Pull Request template to aid you and our reviewers.  Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the xportr codebase remains robust and consistent.  

### The scope of `{xportr}`

`{xportr}`'s scope is to enable R users to write out submission compliant `xpt` files that can be delivered to a Health Authority or to downstream validation software programs.  We see labels, lengths, types, ordering and formats from a dataset specification object (SDTM and ADaM) as being our primary focus. We also see messaging and warnings to users around applying information from the specification file as a primary focus. Please make sure your Pull Request meets this **scope of {xportr}**. If your Pull Request moves beyond this scope, please get in touch with the `{xportr}` team on [slack](https://pharmaverse.slack.com/archives/C030EB2M4GM) or create an issue to discuss.

Please check off each task box as an acknowledgment that you completed the task. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `devel` branch until you have checked off each task.

### Changes Description

* Add a function, `xportr_metadata()`, which allows a user to explicitly set metadata and domain in a single step in the pipeline
* Write several tests, which are mostly copies of previous tests, re-written to use the new "single set metadata" pattern - Not sure how these fit in with other improved test coverage
* I also ran `styler::style_pkg()`, which admittedly hit more files than I meant it to. I can walk this portion back if any of the changes are not desirable
* Huge shoutout to @averissimo whose work with `get_domain()` made this feature much smoother

### Task List

- [x] The spirit of xportr is met in your Pull Request
- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Summary of changes filled out in the above Changes Description.  Can be removed or left blank if changes are minor/self-explanatory.
- [x] Check that your Pull Request is targeting the `devel` branch, Pull Requests to `main` should use the [Release Pull Request Template](https://github.com/atorus-research/xportr/tree/94_pr_template/.github/PULL_REQUEST_TEMPLATE)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/).  Use `styler` package and functions to style files accordingly.
- [x] Updated relevant unit tests or have written new unit tests. See our [Wiki](https://github.com/atorus-research/xportr/wiki/Style-Guide-for-Unit-Tests) for conventions used in this package.
- [x] Creation/updated relevant roxygen headers and examples. See our [Wiki](https://github.com/atorus-research/xportr/wiki/Style-Guide-for-Roxygen-Headers) for conventions used in this package.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Run `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new/updated functions occur on the "Reference" page.
- [x] Update NEWS.md if the changes pertain to a user-facing function (i.e. it has an @export tag) or documentation aimed at users (rather than developers)
- [x] Address any updates needed for vignettes and/or templates
- [x] Link the issue Development Panel so that it closes after successful merging. 
- [x] Pat yourself on the back for a job well done!  Much love to your accomplishment!

### CODE REVIEW

**Comment by EeethB** [comment]
Note on building the pkgdown site: I have added the function I wrote to the pkgdown index. However, whether I `build_site()` from this branch or devel, I get this error:
image

So I think the error is not related to the changes I've made

**Comment by vedhav** [comment]
> Note on building the pkgdown site: I have added the function I wrote to the pkgdown index. However, whether I `build_site()` from this branch or devel, I get this error: image
> 
> So I think the error is not related to the changes I've made

Yes, Ethan, you're right. This was broken when some internal functions were created without ignoring them so pkgdown does not throw errors. This is fixed in the PR #110

File: NEWS.md [hunk 1]
```diff
@@ -1,5 +1,6 @@
 # xportr 0.3.0
 * Fixed an issue where xportr_type would overwrite column labels, widths, and "sas.formats"
+* Added function `xportr_metadata()` to explicitly set metadata at the start of a pipeline
```
Comment (bms63) [inline NEWS.md:3]:
```suggestion
* Added function `xportr_metadata()` to explicitly set metadata at the start of a pipeline - this is just example
```

File: R/df_label.R [hunk 2]
```diff
@@ -38,6 +38,15 @@ xportr_df_label <- function(.df, metacore, domain = NULL) {
...
   ## End of common section
 
+  ## Pull out correct metadata
+  if (is.null(metacore)) {
+    if (is.null(attr(.df, "metadata"))) {
+      stop("Metadata must be set with `metacore` or `xportr_metadata()`")
+    } else {
+      metacore <- attr(.df, "metadata")
+    }
+  }
```
Comment (bms63) [inline R/df_label.R:48]: This seems like this

**Review by vedhav — CHANGES_REQUESTED** [review]
Great job at adding a function to set metadata. Added a few minor comments PTAL.

File: README.Rmd [hunk 3]
```diff
@@ -137,6 +137,19 @@ adsl %>%
...
+```{r, message=FALSE, eval=FALSE}
+adsl %>%
+  xportr_metadata(var_spec, "ADSL") %>%
+  xportr_type() %>%
+  xportr_length() %>%
+  xportr_label() %>%
+  xportr_order() %>%
+  xportr_format() %>%
+  xportr_write("adsl.xpt", label = "Subject-Level Analysis Dataset")
+```
```
[the remark this answers is not in the document]
Comment (bms63) [reply README.Rmd:151]: I agree that the first example should be removed and we should highlight this new and way cleaner `xportr_metadata()` function!

File: R/xportr_metadata.R [hunk 4]
```diff
@@ -0,0 +1,39 @@
+#' Set variable specifications and domain
+#'
+#' @param .df An R object with columns that can be coerced
+#' @param metacore Either a data.frame that has the names of all possible columns
+#'   and their types, or a `Metacore` object from the `Metacore` package. Required
+#'   column names are dataset, variables, type
+#' @param domain Name of the dataset. Ex ADAE/DM. This will be used to subset
+#'   the metacore object. If none is passed it is assumed to be the name of the
+#'   dataset passed in `.df`.
```
Comment (bms63) [inline R/xportr_metadata.R:9]: Please refer to wiki on guidance for filling out the parameter arugments and change `metacore` to `metadata.

Also, after the title can we state the purpose of this function and how it helps reduce the need to fill out additional arguments in the xportr functions.

File: R/xportr_metadata.R [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+#'   type = c("numeric", "character", "numeric", "character")
+#' )
+#'
+#' .df <- data.frame(
+#'   Subj = as.character(123, 456, 789),
+#'   Different = c("a", "b", "c"),
+#'   Val = c("1", "2", "3"),
+#'   Param = c("param1", "param2", "param3")
+#' )
+#'
```
Comment (bms63) [inline R/xportr_metadata.R:27]: instead of `.df` can we use `adlb` since it there are params in here, which is a BDS-Findings style of dataset.

File: tests/testthat/test-type.R
Comment (bms63) [inline tests/testthat/test-type.R:1]: could we create a separate `test-metadata.R` file that looks at the interaction of our functions with this new function.  I feel like there is some edge cases we are missing and the tests can help us correct.

**Review by bms63 — DISMISSED** [review]
Awesome work @EeethB!!  I am very stoked for this change!

File: R/messages.R [hunk 6]
```diff
@@ -49,14 +49,14 @@ var_names_log <- function(tidy_names_df, verbose) {
 
   # Message stating any renamed variables each original variable and it's new name
   if (nrow(only_renames) > 0) {
-    walk(only_renames$renamed_msg, ~ xportr_logger(.x, verbose))
+    purrr::walk(only_renames$renamed_msg, ~ xportr_logger(.x, verbose))
```
[the remark this answers is not in the document]
Comment (EeethB) [reply R/messages.R:52]: I'm all for this! I think this is the Tidyverse style guide recommendation as well

File: R/order.R [hunk 7]
```diff
@@ -37,6 +37,14 @@ xportr_order <- function(.df, metacore, domain = NULL, verbose = getOption("xpor
 
   ## End of common section
 
+  if (is.null(metacore)) {
+    if (is.null(attr(.df, "metadata"))) {
+      stop("Metadata must be set with `metacore` or `xportr_metadata()`")
```
[the remark this answers is not in the document]
Comment (EeethB) [reply R/order.R:42]: Added this, though two of them aren't showing up in the `covr::report()` for me. I've verified that they do have tests though

File: tests/testthat/test-type.R
Comment (EeethB) [reply tests/testthat/test-type.R:1]: I added some more unit tests in `test-metadata.R` to demonstrate equivalence between using `xportr_metadata()` or not. If there are particular use cases you're worried about, I can check them out too

File: tests/testthat/test-label.R [hunk 8]
```diff
@@ -0,0 +1,13 @@
+test_that("xportr_label: error when metadata is not set", {
+  df <- data.frame(
+    Subj = as.character(123, 456, 789),
+    Different = c("a", "b", "c"),
+    Val = c("1", "2", "3"),
+    Param = c("param1", "param2", "param3")
+  )
+
+  expect_error(
+    xportr_type(df),
```
Comment (vedhav) [inline tests/testthat/test-label.R:10]:
```suggestion
    xportr_label(df),
```

Just a typo! For a second I was wondering why was `covr` lying to us 😄

File: tests/testthat/test-order.R [hunk 9]
```diff
@@ -91,6 +91,15 @@ test_that("xportr_order: Expect error if domain is not a character", {
   expect_error(xportr_order(df, df_meta, domain = 1, verbose = "none"))
 })
 
+test_that("xportr_order: error when metadata is not set", {
+  df <- data.frame(c = 1:5, a = "a", d = 5:1, b = LETTERS[1:5])
+
+  expect_error(
+    xportr_type(df),
```
Comment (vedhav) [inline tests/testthat/test-order.R:98]:
```suggestion
    xportr_order(df),
```

File: tests/testthat/test-label.R (hunk 8 above)
Comment (EeethB) [reply tests/testthat/test-label.R:10]: 🤦 Thank you!

File: R/df_label.R [hunk 10]
```diff
@@ -38,6 +38,15 @@ xportr_df_label <- function(.df, metacore, domain = NULL) {
 
   ## End of common section
 
+  ## Pull out correct metadata
+  if (is.null(metacore)) {
+    if (is.null(attr(.df, "metadata"))) {
+      stop("Metadata must be set with `metacore` or `xportr_metadata()`")
```
[the remark this answers is not in the document]
Comment (EeethB) [reply R/df_label.R]: Good idea

**Review by kaz462 — APPROVED** [review]
@EeethB Thanks for the update, LGTM!

File: R/metadata.R [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+#' )
+#'
+#' adlb <- data.frame(
+#'   Subj = as.character(123, 456, 789),
+#'   Different = c("a", "b", "c"),
+#'   Val = c("1", "2", "3"),
+#'   Param = c("param1", "param2", "param3")
+#' )
+#'
+#' xportr_metadata(adlb, metadata, "test")
```
Comment (bms63) [inline R/metadata.R:33]:
```suggestion
#'  adlb %>% 
  xportr_metadata(metadata, "test") %>%
  xportr_type() %>%
  xportr_order() 
```

File: README.Rmd [hunk 12]
```diff
@@ -129,11 +129,12 @@ Each `xportr_` function has been written in a way to take in a part of the speci
...
-  xportr_label(var_spec, "ADSL") %>%
-  xportr_order(var_spec, "ADSL") %>%
-  xportr_format(var_spec, "ADSL") %>%
+  xportr_metadata(var_spec, "ADSL") %>%
+  xportr_type() %>%
+  xportr_length() %>%
+  xportr_label() %>%
+  xportr_order() %>%
+  xportr_format() %>%
   xportr_write("adsl.xpt", label = "Subject-Level Analysis Dataset")
```
Comment (bms63) [inline README.Rmd:138]: @atorus-research/xportr-development-team I'm wondering is there a way to have xportr_write take information from the spec files and apply the dataset label and create the `.xpt` based on the name of the final dataset file with `xportr_metadata`.  Looks so clean like below.  If so, I think this shoudl go into release 0.4

```
adsl %>%
  xportr_metadata(var_spec, "ADSL") %>%
  xportr_type() %>%
  xportr_length() %>%
  xportr_label() %>%
  xportr_order() %>%
  xportr_format() %>%
  xportr_write()
```

Comment (bms63) [reply README.Rmd:138]: AH man --- now I'm thinking ... what an `xportr_appply_all()` that takes all 6 functions and does it all in one swoop??

**Review by bms63 — COMMENTED** [review]
Just a couple of minor updates.  Some bigger ideas to discuss at the Tuesday meeting

File: README.Rmd (hunk 12 above)
Comment (elimillera) [reply README.Rmd:138]: @bms63 We do have xportr_df_label for the dataset label. I think updating the file to <_df_arg_>.xpt would be a nice default. `xportr_metadata_apply` would also be a nice add.

File: R/type.R [hunk 13]
```diff
@@ -67,16 +75,7 @@ xportr_type <- function(.df, metacore, domain = NULL,
...
-    mutate(
-      # _character is used here as a mask of character, in case someone doesn't
-      # want 'character' coerced to character
-      type.x = if_else(type.x %in% characterTypes, "_character", type.x),
-      type.x = if_else(type.x %in% numericTypes, "_numeric", type.x),
-      type.y = tolower(type.y),
-      type.y = if_else(type.y %in% characterTypes, "_character", type.y),
-      type.y = if_else(type.y %in% numericTypes, "_numeric", type.y)
-    )
+  )
```
Comment (elimillera) [inline R/type.R:78]: @EeethB Were these changes causing an issue? Or was it just a merge that took these out?

Comment (EeethB) [reply R/type.R:78]: I think that was just a merge, sorry about that. I will put them back in

**Review by averissimo — APPROVED** [review]
In general it looks good to me. I added some minor comments with suggestions.

File: R/df_label.R [hunk 14]
```diff
@@ -38,6 +38,15 @@ xportr_df_label <- function(.df, metacore, domain = NULL) {
 
   ## End of common section
 
+  ## Pull out correct metadata
+  if (is.null(metacore)) {
+    if (is.null(attr(.df, "metadata"))) {
+      rlang::abort("Metadata must be set with `metacore` or `xportr_metadata()`")
+    } else {
+      metacore <- attr(.df, "metadata")
```
[the remark this answers is not in the document]
Comment (EeethB) [reply R/df_label.R]: That's a very good point, and something I meant to come back and change! It's been updated to match the other attributes' convention

File: R/df_label.R [hunk 15]
```diff
@@ -38,6 +38,15 @@ xportr_df_label <- function(.df, metacore, domain = NULL) {
...
+    }
+  }
```
[the remark this answers is not in the document]
Comment (EeethB) [reply R/df_label.R]: Works for me! I updated all files

**Review by cpiraux — APPROVED** [review]

**Review by vedhav — APPROVED** [review]

**Comment by vedhav** [comment]
> @bms63 @vedhav I think everything is addressed, but it's saying that merging is blocked. Is there any kind of approval you can give on here to unblock it that you know of?

Okay, I think since Ben and I requested changes, we'd have to approve it explicitly. I just approved. And I think you've already addressed the changes requested by Ben so I will dismiss Ben's review to unblock the merge as he's OOO

File: R/xportr_metadata.R [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+  ## Common section to detect domain from argument or pipes
+
+  df_arg <- tryCatch(as_name(enexpr(.df)), error = function(err) NULL)
+  domain <- get_domain(.df, df_arg, domain)
+  if (!is.null(domain)) attr(.df, "_xportr.df_arg_") <- domain
+
+  ## End of common section
+
+  structure(.df, metadata = metacore)
+}
```
[the remark this answers is not in the document]
Comment (cpiraux) [reply R/xportr_metadata.R]: Yes, sure! Thanks @EeethB
